### ISSUE TITLE
Calling RegisterDecorator on short-lived containers causes memory-leak

### ISSUE THREAD

**Issue by torerikh** [simpleinjector/SimpleInjector#539]
It seems like calling verify on short-lived containers results in a memory leak due to dynamicly generated code which is not freed just by disposing the container.

Ideally the verify method should also work according to the EnableDynamicAssemblyCompilation option.

**Comment by dotnetjunkie** [comment]
Could you present a reproducible code sample?

**Comment by dotnetjunkie** [comment]
Ouch. I've been able to reproduce this. Memory starts leaking whenever you use `RegisterDecorator`. This is a regression. This was a bug that was fixed a long time ago, but seems to be reintroduced.

The following program reproduces the issue:

``` c#
private static void Main()
{
    while (true)
    {
        BuildContainer();
        Console.WriteLine(GC.GetTotalMemory(forceFullCollection: true));
    }
}

private static void BuildContainer()
{
    var container = new Container();
    container.Options.EnableDynamicAssemblyCompilation = false;
    container.Register<IProductRepository, SqlPrductRepository>();
    container.RegisterDecorator<IProductRepository, LoggingProductRepository>();          
    container.Verify();
}
```

While this is something that _must_ be fixed, please note that -even without the presence of this bug- you should not create a large number of short-lived container instances. There is a large amount of overhead in the creation of container instances and building their dependency graphs, as stated in [the documentation](https://simpleinjector.org/using):

> **Warning**: Do not create an infinite number of Container instances (such as one instance per request). Doing so will drain the performance of your application. The library is optimized for using a very limited number of Container instances. Creating and initializing Container instances has a large overhead, but resolving from the Container is extremely fast once initialized.

**Comment by dotnetjunkie** [comment]
I published patch releases v4.1.2 and v4.2.2 that fix this bug.

Do note that you will have to call `Container.Dispose()` to prevent memory leaking from happening. A future release will prevent any leakage, even if you omit a call to `Dispose`.
